Add the VMM and page mapping

DrDeano opened the pull request:
This adds the VMM to kmain_64 along with the ability to map and unmap paged.

This uses level 4 paging, level 5 can be an option in the future.

Paging isn't initialised into the CR3 register yet as this is in the arch init phase.

Part of #169

SamTebbs33 commented on src/kernel/arch/x86_64/paging.zig:
```diff
@@ -0,0 +1,806 @@
+}
+
+///
+/// Set or clean the attributes from the VMM attributes to the page entry attributes.
```
clean -> clear

SamTebbs33 commented on src/kernel/arch/x86_64/paging.zig:
```diff
@@ -0,0 +1,806 @@
+}
+
+///
+/// Unmap a the address range from start to end from the Page Table.
```
a the -> an

SamTebbs33 commented on src/kernel/arch/x86_64/paging.zig:
```diff
@@ -0,0 +1,806 @@
+}
+
+///
+/// Unmap a the address range from start to end from the Page Directory Pointer Table.
```
a the -> an

SamTebbs33 commented on src/kernel/arch/x86_64/paging.zig:
```diff
@@ -0,0 +1,806 @@
+        val |= attr;
+    }
+    for (attrs) |attr| {
+        expectEqual(hasAttribute(val, attr), true);
```
This could just be `expect(..)`

SamTebbs33 reviewed (changes requested):
Nice one. This looks a hell of a lot more complex than 32 bit

SamTebbs33 commented:
A post-review idea I had was putting some of the function that operator on certain structs (`unmap*Entry`, `map*Entry`, `*Attribute` etc.) inside those structs for some nice encapsulation.

DrDeano commented:
Could do, add it as an issue. Then can create a new PR for that

SamTebbs33 commented on src/kernel/arch/x86_64/paging.zig:
```diff
@@ -763,7 +763,7 @@ test "map unmap" {
         try unmap(0xFFFFFFFF_10000000, 0xFFFFFFFF_F0000000, allocator, &page_table);
 
-        // PML4 entry 511 and PDPT entry 508 and 511 won't be freed as they are not covered but the freeing range
+        // PML4 entry 511 and PDPT entry 508 and 511 won't be freed as they are not covered buy the freeing range
```
buy -> by ;)

SamTebbs33 commented on src/kernel/arch/x86_64/paging.zig:
```diff
@@ -785,7 +785,7 @@ test "map unmap" {
         try unmap(0xFFFFFFFF_00000000, 0xFFFFFFFF_FFFFF000, allocator, &page_table);
 
-        // PML4 entry 511 and PDPT entry 508 and 511 won't be freed as they are not covered but the freeing range
+        // PML4 entry 511 and PDPT entry 508 and 511 won't be freed as they are not covered buy the freeing range
```
buy -> by ;)

SamTebbs33 reviewed (approved):
Looking gucci

DrDeano committed:
Add the VMM and page mapping

This added the VMM to kmain_64 along with the ability to map and unmap paged.

This uses level 4 paging, level 5 can be an option in the future.

Paging isn't initialised into the CR3 register yet as this is in the arch init phase.

Part #169
